### PR TITLE
Global Interceptor Registration allowed only once

### PR DESCRIPTION
**Pull request by yashykt**
Global Interceptor Registration allowed only once.

This was suggested after the talk on the interception api in the weekly grpc meeting.

### CODE REVIEW

File: src/cpp/client/client_interceptor.cc [hunk 1]
```diff
@@ -28,7 +28,17 @@ experimental::ClientInterceptorFactoryInterface*
...
+  if (internal::g_global_client_interceptor_factory != nullptr) {
+    GPR_ASSERT(false &&
+               "It is illegal to call RegisterGlobalClientInterceptorFactory "
+               "multiple times.");
+  }
   internal::g_global_client_interceptor_factory = factory;
 }
+
+// For testing purposes only.
+void TestOnlyResetGlobalClientInterceptorFactory() {
```
Comment (vjpai) [inline src/cpp/client/client_interceptor.cc]: I think it's ok to declare this in the .h file. The TestOnly bit should make it clear that it's not for use. You could even make it namespace grpc::testing like we use in ServerContextTestSpouse.

File: test/cpp/end2end/client_interceptors_end2end_test.cc [hunk 2]
```diff
@@ -40,6 +40,10 @@
 #include <gtest/gtest.h>
 
 namespace grpc {
+
+namespace experimental {
+void TestOnlyResetGlobalClientInterceptorFactory();
```
Comment (vjpai) [inline test/cpp/end2end/client_interceptors_end2end_test.cc]: Prefer removing this forward decl and declaring in a header file instead.

File: src/cpp/client/client_interceptor.cc (hunk 1 above)
Comment (yashykt) [reply src/cpp/client/client_interceptor.cc]: done

File: test/cpp/end2end/client_interceptors_end2end_test.cc (hunk 2 above)
Comment (yashykt) [reply test/cpp/end2end/client_interceptors_end2end_test.cc]: done

**Comment by yashykt** [comment]
ping :)

**Pull request merged**